helix-term/src/commands/typed.rs

```diff
@@ -334,6 +334,11 @@ fn write_impl(
     let (view, doc) = current!(cx.editor);
     let path = path.map(AsRef::as_ref);
 
+    // `path.is_none()` allows :w <path> for other files
+    if path.is_none() && doc.detect_readonly() && !force {
+        bail!("file is readonly, use ! to overwrite")
```

Comment:
`:w!` won't behave any different than `:w`: if the underlying file is read-only then any write will fail. I don't think we need any extra handling in the write functions for this. The errors when attempting to write to the file will already show up as errors on the statusline.